crates/storage-rocksdb/src/writer.rs

```diff
@@ -0,0 +1,214 @@
+// Copyright (c) 2023 -  Restate Software, Inc., Restate GmbH.
+// All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use crate::DB;
+use futures::ready;
+use futures_util::FutureExt;
+use log::debug;
+use restate_storage_api::StorageError;
+use restate_types::errors::ThreadJoinError;
+use rocksdb::WriteBatch;
+use std::future::Future;
+use std::panic;
+use std::panic::AssertUnwindSafe;
+use std::pin::Pin;
+use std::sync::Arc;
+use std::task::{Context, Poll};
+use tokio::sync::mpsc::{UnboundedReceiver, UnboundedSender};
+use tokio::sync::oneshot::Sender;
+
+pub struct WriteCommand {
+    write_batch: WriteBatch,
+    response_tx: Sender<Result<(), StorageError>>,
+}
+
+impl WriteCommand {
+    pub fn new(write_batch: WriteBatch, response_tx: Sender<Result<(), StorageError>>) -> Self {
+        Self {
+            write_batch,
+            response_tx,
+        }
+    }
+}
+
+type Error = anyhow::Error;
+
+pub struct Writer {
+    db: Arc<DB>,
+    rx: UnboundedReceiver<WriteCommand>,
+
+    // for creating `WriterHandle`s
+    tx: UnboundedSender<WriteCommand>,
+}
+
+impl Writer {
+    pub fn new(db: Arc<DB>) -> Self {
+        let (tx, rx) = tokio::sync::mpsc::unbounded_channel();
+        Self { db, rx, tx }
+    }
+
+    pub fn create_writer_handle(&self) -> WriterHandle {
+        WriterHandle {
+            tx: self.tx.clone(),
+        }
+    }
+
+    pub fn run(self, shutdown_watch: drain::Watch) -> JoinHandle {
+        let (tx, rx) = tokio::sync::oneshot::channel();
+
+        std::thread::Builder::new()
+            .name("RocksDB writer thread".to_owned())
+            .spawn(|| {
+                // AssertUnwindSafe is safe because we don't access self after catch_unwind again
+                let result = panic::catch_unwind(AssertUnwindSafe(|| {
```

Comment:
Just for future reference: The idea was to use a dedicated thread to avoid blocking the `spawn_blocking` thread pool. The reason for creating a new Tokio runtime is to be able to react to the shutdown signal (being able to select over multiple inputs). W/o the shutdown signal, the tests failed with a segfault.